File: packages/sdk-java/harness-sdk-guardrails/src/main/java/com/harness/guardrails/ChineseNameRecognizer.java
package com.harness.guardrails;

import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Chinese Name Recognizer.
 *
 * Provides Chinese name detection using surname lists and rules.
 * Supports both single-character and compound surnames.
 *
 * Example:
 * <pre>
 * ChineseNameRecognizer recognizer = new ChineseNameRecognizer();
 * List&lt;NameMatch&gt; names = recognizer.recognize("联系人：张伟，电话13912345678");
 * // [NameMatch(text='张伟', surname='张', given_name='伟', score=0.9)]
 * </pre>
 */
public class ChineseNameRecognizer {

    // Common Chinese surnames (top 100, covering ~85% of population)
    private static final Set<Character> COMMON_SURNAMES = Set.of(
        '王', '李', '张', '刘', '陈', '杨', '赵', '黄', '周', '吴',
        '徐', '孙', '胡', '朱', '高', '林', '何', '郭', '马', '罗',
        '梁', '宋', '郑', '谢', '韩', '唐', '冯', '于', '董', '萧',
        '程', '曹', '袁', '邓', '许', '傅', '沈', '曾', '彭', '吕',
        '苏', '卢', '蒋', '蔡', '贾', '丁', '魏', '薛', '叶', '阎',
        '余', '潘', '杜', '戴', '夏', '钟', '汪', '田', '任', '姜',
        '范', '方', '石', '姚', '谭', '廖', '邹', '熊', '金', '陆',
        '郝', '孔', '白', '崔', '康', '毛', '邱', '秦', '江', '史',
        '顾', '侯', '邵', '孟', '龙', '万', '段', '曹', '钱', '汤',
        '尹', '黎', '易', '常', '武', '乔', '贺', '赖', '龚', '文'
    );

    // Compound surnames (two-character surnames)
    private static final List<String> COMPOUND_SURNAMES = List.of(
        "欧阳", "上官", "司马", "诸葛", "东方", "皇甫", "尉迟", "公孙", "令狐", "宇文",
        "长孙", "慕容", "司徒", "南宫", "独孤", "百里", "端木", "轩辕", "赫连", "澹台"
    );

    // False positives to filter out
    private static final Set<String> FALSE_POSITIVES = Set.of(
        // Common non-name phrases
        "手机号", "电话号", "身份证", "银行卡", "邮箱地", "联系方",
        "公司名", "产品名", "用户名", "账号", "账户",
        // City names
        "北京市", "上海市", "广州市", "深圳市",
        // Days of week
        "周一", "周二", "周三", "周四", "周五", "周六", "周日",
        // Months
        "一月", "二月", "三月", "四月", "五月", "六月",
        "七月", "八月", "九月", "十月", "十一月", "十二月"
    );

    // Common given name characters
    private static final Set<Character> COMMON_GIVEN_CHARS = Set.of(
        '伟', '芳', '娜', '敏', '静', '丽', '强', '磊', '军', '洋',
        '勇', '艳', '杰', '娟', '涛', '明', '超', '秀', '霞', '平',
        '刚', '桂', '英', '兰', '华', '建', '国', '文', '辉', '斌',
        '波', '宇', '红', '梅', '玲', '鹏', '峰', '毅', '浩', '清',
        '云', '翔', '林', '海', '天', '山', '风', '龙', '飞'
    );

    // Context keywords that indicate names
    private static final Set<String> NAME_CONTEXT = Set.of(
        "姓名", "名字", "联系人", "联系", "客户", "用户", "本人", "持卡人",
        "叫", "是", "叫作", "叫做"
    );

    // Patterns for name detection
    private static final Pattern[] NAME_PATTERNS = {
        // Context keyword + name
        Pattern.compile("(?:姓名|名字|联系人|联系|客户|用户|本人|持卡人|叫|是)[:：]?\\s*([\\u4e00-\\u9fff]{2,4})"),
        // Name + possessive + other info
        Pattern.compile("([\\u4e00-\\u9fff]{2,3})的(?:手机|电话|邮箱|身份证|银行卡)"),
        // Standalone 2-4 character Chinese name
        Pattern.compile("(?<![\\u4e00-\\u9fff])([\\u4e00-\\u9fff]{2,4})(?![\\u4e00-\\u9fff])")
    };

    private final double minScore;

    /**
     * Create a Chinese name recognizer with default settings.
     */
    public ChineseNameRecognizer() {
        this(0.5);
    }

    /**
     * Create a Chinese name recognizer.
     *
     * @param minScore Minimum confidence threshold
     */
    public ChineseNameRecognizer(double minScore) {
        this.minScore = minScore;
    }

    /**
     * Recognize Chinese names in text.
     *
     * @param text Text to analyze
     * @return List of detected name matches
     */
    public List<NameMatch> recognize(String text) {
        List<NameMatch> results = new ArrayList<>();

        for (Pattern pattern : NAME_PATTERNS) {
            Matcher matcher = pattern.matcher(text);
            while (matcher.find()) {
                String candidate = matcher.group(1);
                int start = matcher.start(1);
                int end = matcher.end(1);

                NameMatch nameMatch = validateName(candidate, start, end, text);
                if (nameMatch != null && nameMatch.score >= minScore) {
                    // Check for duplicates
                    if (!isDuplicate(results, nameMatch)) {
                        results.add(nameMatch);
                    }
                }
            }
        }

        return results;
    }

    /**
     * Detect Chinese names and return as PII entities.
     *
     * @param text Text to analyze
     * @return List of PII entities
     */
    public List<PIIEntity> detect(String text) {
        List<PIIEntity> entities = new ArrayList<>();
        List<NameMatch> names = recognize(text);

        for (NameMatch name : names) {
            entities.add(new PIIEntity(
                PIIEntity.Type.NAME,
                name.text,
                name.start,
                name.end,
                name.score
            ));
        }

        return entities;
    }

    /**
     * Validate if a candidate string is a valid Chinese name.
     */
    private NameMatch validateName(String candidate, int start, int end, String fullText) {
        if (candidate == null || candidate.isEmpty()) {
            return null;
        }

        // Length check (2-4 characters)
        if (candidate.length() < 2 || candidate.length() > 4) {
            return null;
        }

        // Must be all Chinese characters
        for (char c : candidate.toCharArray()) {
            if (c < '\u4e00' || c > '\u9fff') {
                return null;
            }
        }

        // Check for false positives
        if (FALSE_POSITIVES.contains(candidate)) {
            return null;
        }

        // Check surname
        String surname = null;
        String givenName = null;

        // Check compound surnames first
        for (String cs : COMPOUND_SURNAMES) {
            if (candidate.startsWith(cs)) {
                surname = cs;
                givenName = candidate.substring(cs.length());
                break;
            }
        }

        // Check single-character surnames
        if (surname == null) {
            char firstChar = candidate.charAt(0);
            if (COMMON_SURNAMES.contains(firstChar)) {
                surname = String.valueOf(firstChar);
                givenName = candidate.substring(1);
            }
        }

        if (surname == null) {
            return null;
        }

        // Calculate score
        double score = calculateScore(candidate, surname, givenName, fullText, start);

        return new NameMatch(candidate, start, end, score, surname, givenName);
    }

    /**
     * Calculate confidence score for a name.
     */
    private double calculateScore(String candidate, String surname, String givenName,
                                   String fullText, int matchStart) {
        double score = 0.7;

        // Bonus for common surname
        if (COMMON_SURNAMES.contains(surname.charAt(0)) || COMPOUND_SURNAMES.contains(surname)) {
            score += 0.1;
        }

        // Bonus for reasonable given name length (1-2 chars)
        if (givenName != null && givenName.length() >= 1 && givenName.length() <= 2) {
            score += 0.1;
        }

        // Bonus for common given name characters
        if (givenName != null) {
            boolean allCommon = true;
            for (char c : givenName.toCharArray()) {
                if (!COMMON_GIVEN_CHARS.contains(c)) {
                    allCommon = false;
                    break;
                }
            }
            if (allCommon) {
                score += 0.05;
            }
        }

        // Bonus for context keywords
        if (hasContext(fullText, matchStart, NAME_CONTEXT)) {
            score += 0.1;
        }

        return Math.min(score, 1.0);
    }

    /**
     * Check if context keywords appear within 30 characters before the match.
     */
    private boolean hasContext(String text, int matchStart, Set<String> contextKeywords) {
        int windowStart = Math.max(0, matchStart - 30);
        String contextWindow = text.substring(windowStart, matchStart).toLowerCase();

        for (String keyword : contextKeywords) {
            if (contextWindow.contains(keyword.toLowerCase())) {
                return true;
            }
        }

        return false;
    }

    /**
     * Check if a match is a duplicate of existing results.
     */
    private boolean isDuplicate(List<NameMatch> results, NameMatch newMatch) {
        for (NameMatch existing : results) {
            // Consider it a duplicate if there's significant overlap
            if (newMatch.start < existing.end && newMatch.end > existing.start) {
                return true;
            }
        }
        return false;
    }

    /**
     * Name match result.
     */
    public static class NameMatch {
        public final String text;
        public final int start;
        public final int end;
        public final double score;
        public final String surname;
        public final String givenName;

        public NameMatch(String text, int start, int end, double score, String surname, String givenName) {
            this.text = text;
            this.start = start;
            this.end = end;
            this.score = score;
            this.surname = surname;
            this.givenName = givenName;
        }

        @Override
        public String toString() {
            return String.format("NameMatch{text='%s', surname='%s', given='%s', score=%.2f}",
                text, surname, givenName, score);
        }
    }
}

File: packages/sdk/src/harness/guardrails/chinese_name_recognizer.py
"""
中文姓名识别器

提供多种中文姓名识别方案：
1. 基于姓氏列表 + 规则
2. 基于 spaCy 中文 NER + 姓氏过滤
3. 混合方案（推荐）
"""

import re
from dataclasses import dataclass

# 中国常见姓氏（前100大姓，覆盖约85%人口）
COMMON_SURNAMES = set(
    "王李张刘陈杨赵黄周吴徐孙胡朱高林何郭马罗梁宋郑谢韩唐冯于董萧程曹袁邓许傅沈曾彭吕苏卢蒋蔡贾丁魏薛叶阎余潘杜戴夏钟汪田任姜范方石姚谭廖邹熊金陆郝孔白崔康毛邱秦江史顾侯邵孟龙万段漕钱汤尹黎易常武乔贺赖龚文"
)

# 复姓
COMPOUND_SURNAMES = [
    "欧阳",
    "上官",
    "司马",
    "诸葛",
    "东方",
    "皇甫",
    "尉迟",
    "公孙",
    "令狐",
    "宇文",
    "长孙",
    "慕容",
    "司徒",
    "南宫",
    "独孤",
    "百里",
    "端木",
    "轩辕",
    "赫连",
    "澹台",
]


@dataclass
class NameMatch:
    """姓名匹配结果"""

    text: str
    start: int
    end: int
    score: float
    surname: str
    given_name: str


class ChineseNameRecognizer:
    """
    中文姓名识别器

    使用方法:
        recognizer = ChineseNameRecognizer()

        names = recognizer.recognize("联系人：张伟，电话13912345678")
        # [NameMatch(text='张伟', surname='张', given_name='伟', score=0.9)]
    """

    def __init__(self, use_spacy: bool = True, min_score: float = 0.5):
        """
        初始化

        Args:
            use_spacy: 是否使用 spaCy NER 辅助
            min_score: 最小置信度阈值
        """
        self.use_spacy = use_spacy
        self.min_score = min_score
        self.nlp = None

        if use_spacy:
            try:
                import spacy

                self.nlp = spacy.load("zh_core_web_sm")
            except Exception:
                self.nlp = None

    def recognize(self, text: str) -> list[NameMatch]:
        """
        识别文本中的中文姓名

        Args:
            text: 待识别文本

        Returns:
            识别到的姓名列表
        """
        results = []

        # 方案1：基于规则识别
        rule_results = self._recognize_by_rules(text)
        results.extend(rule_results)

        # 方案2：spaCy NER + 姓氏过滤
        if self.nlp:
            spacy_results = self._recognize_by_spacy(text)
            # 合并结果，去重
            results = self._merge_results(results, spacy_results)

        # 过滤低分结果
        results = [r for r in results if r.score >= self.min_score]

        return results

    def _recognize_by_rules(self, text: str) -> list[NameMatch]:
        """基于规则识别姓名"""
        results = []

        # 模式1：姓名上下文关键词
        name_patterns = [
            # "叫/是/姓名" + 姓名
            r"(?:叫|是|姓名|名字|联系人|联系|客户|用户|本人|持卡人)[:：]?\s*([一-龥]{2,4})",
            # 姓名 + "的" + 其他
            r"([一-龥]{2,3})的(?:手机|电话|邮箱|身份证|银行卡)",
            # 姓氏开头的2-4字姓名
            r"(?<![一-龥])([一-龥]{2,4})(?![一-龥])",
        ]

        for pattern in name_patterns:
            matches = re.finditer(pattern, text)
            for match in matches:
                candidate = match.group(1)
                start, end = match.start(1), match.end(1)

                # 验证是否为有效姓名
                name_match = self._validate_name(candidate, start, end)
                if name_match:
                    results.append(name_match)

        return results

    def _recognize_by_spacy(self, text: str) -> list[NameMatch]:
        """使用 spaCy NER 识别"""
        results = []

        try:
            doc = self.nlp(text)
            for ent in doc.ents:
                if ent.label_ == "PERSON":
                    candidate = ent.text
                    start, end = ent.start_char, ent.end_char

                    # 姓氏过滤
                    name_match = self._validate_name(candidate, start, end)
                    if name_match:
                        results.append(name_match)
        except Exception:
            pass

        return results

    def _validate_name(self, candidate: str, start: int, end: int) -> NameMatch | None:
        """
        验证候选字符串是否为有效中文姓名

        有效姓名条件：
        1. 长度 2-4 个汉字
        2. 以常见姓氏开头
        3. 不包含常见非姓名词汇
        """
        if not candidate:
            return None

        # 长度检查
        if len(candidate) < 2 or len(candidate) > 4:
            return None

        # 必须全是汉字
        if not all("\u4e00" <= c <= "\u9fff" for c in candidate):
            return None

        # 检查姓氏
        surname = None
        given_name = None

        # 检查复姓
        for cs in COMPOUND_SURNAMES:
            if candidate.startswith(cs):
                surname = cs
                given_name = candidate[len(cs) :]
                break

        # 检查单姓
        if not surname:
            first_char = candidate[0]
            if first_char in COMMON_SURNAMES:
                surname = first_char
                given_name = candidate[1:]

        if not surname:
            return None

        # 过滤常见误报词汇
        false_positives = {
            "手机号",
            "电话号",
            "身份证",
            "银行卡",
            "邮箱地",
            "联系方",
            "公司名",
            "产品名",
            "用户名",
            "账号",
            "账户",
            "北京市",
            "上海市",
            "广州市",
            "深圳市",  # 城市名
            "周一",
            "周二",
            "周三",
            "周四",
            "周五",
            "周六",
            "周日",  # 星期
            "一月",
            "二月",
            "三月",
            "四月",
            "五月",
            "六月",  # 月份
        }

        if candidate in false_positives:
            return None

        # 计算置信度
        score = self._calculate_score(candidate, surname, given_name)

        return NameMatch(
            text=candidate,
            start=start,
            end=end,
            score=score,
            surname=surname,
            given_name=given_name,
        )

    def _calculate_score(self, candidate: str, surname: str, given_name: str) -> float:
        """
        计算姓名置信度

        评分规则：
        - 基础分：0.7
        - 姓氏在常见姓氏中：+0.1
        - 名字长度 1-2 字：+0.1
        - 名字不包含罕见字：+0.05
        - 被 spaCy 识别为人名：+0.1（在 merge 时处理）
        """
        score = 0.7

        # 常见姓氏加分
        if surname in COMMON_SURNAMES or surname in COMPOUND_SURNAMES:
            score += 0.1

        # 名字长度合理
        if given_name and 1 <= len(given_name) <= 2:
            score += 0.1

        # 名字不含罕见字（简单判断）
        common_chars = set(
            "伟芳娜敏静丽强磊军洋勇艳杰娟涛明超秀霞平刚桂英兰华建国文辉斌波宇红梅玲鹏峰毅浩清云翔林海天山风龙飞"
        )
        if given_name and all(c in common_chars for c in given_name):
            score += 0.05

        return min(score, 1.0)

    def _merge_results(
        self, rule_results: list[NameMatch], spacy_results: list[NameMatch]
    ) -> list[NameMatch]:
        """合并规则和 spaCy 结果"""
        # 使用位置作为唯一标识
        seen = {}
        for r in rule_results:
            seen[(r.start, r.end)] = r

        for r in spacy_results:
            key = (r.start, r.end)
            if key in seen:
                # 如果重叠，选择分数更高的
                if r.score > seen[key].score:
                    seen[key] = r
            else:
                # spaCy 独有结果，加分
                r.score = min(r.score + 0.1, 1.0)
                seen[key] = r

        return list(seen.values())


def create_name_recognizer(use_spacy: bool = True) -> ChineseNameRecognizer:
    """创建中文姓名识别器"""
    return ChineseNameRecognizer(use_spacy=use_spacy)


# 便捷函数
def extract_chinese_names(text: str) -> list[str]:
    """
    快速提取文本中的中文姓名

    使用方法:
        names = extract_chinese_names("联系人：张伟，电话13912345678")
        # ['张伟']
    """
    recognizer = ChineseNameRecognizer()
    matches = recognizer.recognize(text)
    return [m.text for m in matches]
